src/core/mesh/qgsmesheditor.cpp

```diff
@@ -0,0 +1,1314 @@
+    bool isTooClose = false;
+    int triangleIndex = mTriangularMesh->faceIndexForPoint_v2( pointInTriangularMesh );
+    if ( triangleIndex != -1 )
+    {
+      const QgsMeshFace face = mTriangularMesh->triangles().at( triangleIndex );
+      for ( int j = 0; j < 3; ++j )
+      {
+        const QgsPointXY &facePoint = mTriangularMesh->vertices().at( face.at( j ) );
+        double dist = pointInTriangularMesh.distance( facePoint );
+        if ( dist < tolerance )
+        {
+          isTooClose = true;
+          break;
+        }
+      }
+    }
+
+    if ( !isTooClose )
+      verticesInLayerCoordinate[i] = mTriangularMesh->triangularToNativeCoordinates( vertices.at( i ) );
+    else
+      verticesInLayerCoordinate[i] = QgsMeshVertex();
+
+    if ( verticesInLayerCoordinate.at( i ).isEmpty() )
+      ignoredVertex++;
+  }
+
+  if ( ignoredVertex < vertices.count() )
+  {
+    mUndoStack->push( new QgsMeshLayerUndoCommandAddVertices( this, verticesInLayerCoordinate ) );
+    emit meshEdited();
+  }
+
+  return vertices.count() - ignoredVertex;
+}
+
+int QgsMeshEditor::addPointsAsVertices( const QVector<QgsPoint> &point, double tolerance )
+{
+  return addVertices( point, tolerance );
+}
+
+QgsTopologicMesh::TopologicError QgsMeshEditor::removeVertices( const QList<int> &verticesToRemoveIndexes, bool fillHoles )
+{
+  QgsTopologicMesh::TopologicError error;
+  if ( !fillHoles )
+  {
+    QSet<int> concernedNativeFaces;
+    for ( const int vi : verticesToRemoveIndexes )
+      concernedNativeFaces.unite( mTopologicMesh.facesAroundVertex( vi ).toSet() );
+
+    error = mTopologicMesh.areFacesCanBeRemoved( concernedNativeFaces.values() );
+    if ( error.errorType != QgsTopologicMesh::NoError )
+      return error;
+  }
+
+  mUndoStack->push( new QgsMeshLayerUndoCommandRemoveVertices( this, verticesToRemoveIndexes, fillHoles ) );
+
+  return error;
+}
+
+void QgsMeshEditor::stopEdition()
+{
+  mTopologicMesh.reIndex();
+}
+
+QgsMeshEditor::~QgsMeshEditor() = default;
+
+QgsTopologicMesh QgsTopologicMesh::createTopologicMesh( QgsMesh *mesh, TopologicError &error )
+{
+  QgsTopologicMesh topologicMesh;
+  topologicMesh.mMesh = mesh;
+  topologicMesh.mVertexToface = QVector<int>( mesh->vertexCount(), -1 );
+  error.errorType = NoError;
+
+  for ( int i = 0; i < mesh->faceCount(); ++i )
+  {
+    error = counterClockWiseFaces( mesh->faces[i], mesh );
+    if ( error.errorType != NoError )
+    {
+      if ( error.errorType == InvalidFace || error.errorType == FlatFace )
+        error.elementIndex = i;
+      break;
+    }
+  }
+
+  if ( error.errorType == NoError )
+  {
+    TopologicFaces subMesh = topologicMesh.createTopologicFaces( mesh->faces, error, false, true );
+    topologicMesh.mFacesNeighborhood = subMesh.mFacesNeighborhood;
+  }
+
+  return topologicMesh;
+}
+
+QgsTopologicMesh::TopologicFaces QgsTopologicMesh::createNewTopologicFaces( const QVector<QgsMeshFace> &faces, QgsTopologicMesh::TopologicError &error )
```

Comment:
I think you need "Topological" instead of "Topologic", not sure if "Topologic" is English word.

https://en.wikipedia.org/wiki/Topology